numpy/typing/__init__.py

```diff
@@ -140,6 +140,20 @@
 The `~numpy.timedelta64` class is not considered a subclass of `~numpy.signedinteger`,
 the former only inheriting from `~numpy.generic` while static type checking.
 
+0D arrays
+~~~~~~~~~
+
+During runtime numpy aggressively casts any passed 0D arrays into their
+corresponding `~numpy.generic` instance. Until the introduction of shape
+typing (see :pep:`646`) it is unfortunately not possible to make the
+necessary distinction between 0D and >0D arrays. While thus not strictly
+correct, all operations are that can potentially perform a 0D-array -> scalar
+cast are currently annotated as exclusively returning an `ndarray`.
+
+If it is known in advance that an operation _will_ perform a
+0D-array -> scalar cast, then one can consider manually remedying the
+situation with either `typing.cast` or a ``# type: ignore`` comment.
+
```

Comment:
TLDR: The new return annotations will now just be somewhat inconvenient for 0D arrays, rather than 0D _and_ ND arrays.

Reply:
Also a second question: are there any parts of the numpy documentation that describe 0D arrays and/or numpy's aggressive 0D-to-scalar casting? If so, then it might be useful the place a link here.

Reply:
The casts are scattered through the C code as `PyArray_Return`. The policy, such as it is, is documented as
```
.. c:function:: PyObject* PyArray_Return(PyArrayObject* arr)

    This function steals a reference to *arr*.

    This function checks to see if *arr* is a 0-dimensional array and,
    if so, returns the appropriate array scalar. It should be used
    whenever 0-dimensional arrays could be returned to Python.
```